File: lib/rag_engine.py
"""
RAG (Retrieval-Augmented Generation) 엔진

KeywordPulse의 키워드 분석 결과를 자연어 텍스트로 변환하는 RAG 시스템 핵심 모듈입니다.
"""
from typing import List, Dict, Any, Optional
import time
from lib.logger import get_logger

# 로거 인스턴스 생성
logger = get_logger("rag_engine")

def generate_analysis_text(keywords: List[Dict[str, Any]]) -> str:
    """
    입력된 키워드 목록을 기반으로 분석 텍스트를 생성합니다.
    
    Args:
        keywords: 키워드 정보가 담긴 사전 리스트
                  [{"keyword": "AI 마케팅", "monthlySearches": 25000, "competitionRate": 0.25, "score": 85}, ...]
    
    Returns:
        str: 마크다운 형식의 분석 텍스트
    """
    # 시작 시간 기록
    start_time = time.time()
    
    # 로깅: 분석 요청 시작
    logger.info(
        "키워드 분석 텍스트 생성 시작",
        context={"keyword_count": len(keywords)}
    )
    
    try:
        if not keywords:
            logger.warning("분석할 키워드가 없음", context={"keywords": []})
            return "분석할 키워드가 없습니다."
        
        # 점수 기준 상위 키워드 선별
        top_keywords = sorted(keywords, key=lambda x: x.get('score', 0), reverse=True)[:5]
        
        # 로깅: 선별된 상위 키워드
        logger.debug(
            "상위 키워드 선별 완료", 
            context={
                "top_keywords": [kw.get('keyword') for kw in top_keywords],
                "top_scores": [kw.get('score') for kw in top_keywords]
            }
        )
        
        # 분석 텍스트 생성
        summary = ""
        
        # 서두
        if top_keywords:
            summary += f"이번 분석된 키워드 중 '{top_keywords[0]['keyword']}'는 가장 높은 추천 점수를 기록했습니다.\n\n"
        
        # 주요 키워드 목록
        summary += "## 주요 키워드 분석\n\n"
        for kw in top_keywords:
            score = kw.get('score', 0)
            recommendation = ""
            
            if score >= 80:
                recommendation = "🟢 강력 추천"
            elif score >= 50:
                recommendation = "🟡 추천"
            else:
                recommendation = "⚪ 낮은 우선순위"
            
            summary += f"- **{kw['keyword']}**: 검색량 {kw.get('monthlySearches', 0):,}회, "
            summary += f"경쟁률 {kw.get('competitionRate', 0):.2f}, "
            summary += f"점수 {score}점 ({recommendation})\n"
        
        # 결론
        high_score_keywords = [kw for kw in keywords if kw.get('score', 0) >= 80]
        if high_score_keywords:
            summary += "\n## 추천 전략\n\n"
            summary += f"특히 점수가 80점 이상인 {len(high_score_keywords)}개 키워드는 "
            summary += "콘텐츠 제작 우선순위로 고려하시길 권장합니다.\n"
        
        # 처리 시간 계산
        elapsed_time = time.time() - start_time
        
        # 로깅: 분석 요청 완료
        logger.info(
            "키워드 분석 텍스트 생성 완료",
            context={
                "keyword_count": len(keywords),
                "top_keyword": top_keywords[0]['keyword'] if top_keywords else None,
                "high_score_count": len(high_score_keywords),
                "processing_time_ms": round(elapsed_time * 1000, 2)
            }
        )
        
        return summary
    except Exception as e:
        # 로깅: 분석 중 오류
        logger.error(
            "키워드 분석 텍스트 생성 중 오류 발생",
            context={"keyword_count": len(keywords) if keywords else 0},
            error=e
        )
        return "분석 중 오류가 발생했습니다. 다시 시도해 주세요."

# LLM 기반 확장 가능성을 위한 인터페이스
def generate_with_llm(keywords: List[Dict[str, Any]], provider: str = "template") -> str:
    """
    LLM을 활용한 키워드 분석 텍스트 생성 (확장용)
    
    Args:
        keywords: 키워드 정보 리스트
        provider: 'template'(기본값), 'openai', 'claude' 등
    
    Returns:
        str: 생성된 분석 텍스트
    """
    # 로깅: LLM 생성 요청
    logger.info(
        f"LLM 기반 텍스트 생성 요청",
        context={
            "provider": provider,
            "keyword_count": len(keywords) if keywords else 0
        }
    )
    
    try:
        if provider == "template":
            return generate_analysis_text(keywords)
        
        # 향후 다른 프로바이더 지원 시 분기 처리
        if provider == "openai":
            # 현재는 지원되지 않음
            logger.warning(
                "OpenAI LLM 제공자는 아직 구현되지 않음",
                context={"fallback": "template"}
            )
            return generate_analysis_text(keywords)
        
        if provider == "claude":
            # 현재는 지원되지 않음
            logger.warning(
                "Claude LLM 제공자는 아직 구현되지 않음",
                context={"fallback": "template"}
            )
            return generate_analysis_text(keywords)
        
        # 기본 템플릿 사용
        logger.warning(
            f"알 수 없는 LLM 제공자: {provider}",
            context={"fallback": "template"}
        )
        return generate_analysis_text(keywords)
    except Exception as e:
        # 로깅: LLM 생성 중 오류
        logger.error(
            "LLM 기반 텍스트 생성 중 오류 발생",
            context={"provider": provider},
            error=e
        )
        return "LLM 분석 중 오류가 발생했습니다. 기본 템플릿을 사용해 주세요."

# 키워드 카테고리 캐싱
_keyword_category_cache = {}

def categorize_keyword(keyword: str) -> str:
    """
    키워드를 카테고리로 분류합니다.
    
    Args:
        keyword: 분류할 키워드 문자열
        
    Returns:
        str: 카테고리 이름 ('디지털 마케팅', 'AI 기술', '앱 개발' 등)
    """
    # 캐시 확인
    if keyword in _keyword_category_cache:
        logger.debug(
            "키워드 카테고리 캐시 사용",
            context={
                "keyword": keyword,
                "category": _keyword_category_cache[keyword]
            }
        )
        return _keyword_category_cache[keyword]
    
    # 소문자 변환 및 공백 제거
    k = keyword.lower().strip()
    
    # 기본 카테고리
    category = "일반"
    
    # 카테고리 분류 패턴
    patterns = {
        '디지털 마케팅': ['마케팅', 'seo', '광고', '콘텐츠', '퍼포먼스', '인플루언서', '바이럴', '브랜딩'],
        'AI 기술': ['ai', '인공지능', '머신러닝', 'gpt', '딥러닝', '자연어', '신경망', '강화학습'],
        '앱 개발': ['개발', '프로그래밍', '앱', 'app', '코딩', '소프트웨어', '웹', '모바일'],
        '3D 모델링/AI': ['3d', '모델링', '블렌더', 'blender', '렌더링', '애니메이션', 'cg', '그래픽']
    }
    
    # 패턴 매칭
    for cat, keywords in patterns.items():
        if any(kw in k for kw in keywords):
            category = cat
            break
    
    # 캐시 저장
    _keyword_category_cache[keyword] = category
    
    logger.debug(
        "키워드 카테고리 분류 완료",
        context={
            "keyword": keyword,
            "category": category
        }
    )
    
    return category 

File: app/lib/rag_engine.ts
/**
 * RAG (Retrieval Augmented Generation) 엔진 모듈
 * 검색 기반 생성 기능을 제공합니다.
 */

import logger from './logger';

export interface RagOptions {
  maxResults?: number;
  threshold?: number;
  includeMetadata?: boolean;
  searchProvider?: 'supabase' | 'elasticsearch' | 'pinecone' | 'llm';
}

export interface RagDocument {
  id: string;
  content: string;
  metadata?: Record<string, any>;
  score?: number;
}

export interface RagSearchResult {
  query: string;
  results: RagDocument[];
  totalCount: number;
  executionTime: number;
}

export interface RagGenerationResult {
  sourceQuery: string;
  generatedContent: string;
  citations: RagDocument[];
  confidence: number;
}

/**
 * 키워드 카테고리 타입
 */
export type KeywordCategory = 
  | 'AI 기술'
  | '디지털 마케팅'
  | '앱 개발' 
  | '교육/학습'
  | '건강/의료'
  | '금융/투자'
  | '3D 모델링/AI'
  | '일반';

/**
 * 지원되는 언어 타입
 */
export type SupportedLanguage = 'ko' | 'en';

/**
 * 사용자 설정 인터페이스
 */
export interface UserPreferences {
  language?: SupportedLanguage;
  industry?: string;
  insightCount?: number;
  strategyCount?: number;
  detailLevel?: 'basic' | 'detailed' | 'comprehensive';
}

/**
 * RAG 검색을 실행합니다
 * @param query 검색 쿼리
 * @param options 검색 옵션
 * @returns 검색 결과
 */
export async function ragSearch(query: string, options: RagOptions = {}): Promise<RagSearchResult> {
  const startTime = Date.now();
  const {
    maxResults = 5,
    threshold = 0.7,
    includeMetadata = true,
    searchProvider = 'supabase'
  } = options;

  try {
    logger.log({
      message: `RAG 검색 시작: "${query}"`,
      level: 'info',
      context: { maxResults, threshold, searchProvider }
    });

    // 여기서는 더미 데이터 반환
    // 실제 구현에서는 searchProvider에 따라 다른 데이터베이스나 벡터 스토어를 쿼리합니다
    const results: RagDocument[] = [
      {
        id: '1',
        content: '인공지능(AI)은 컴퓨터 시스템이 인간의 지능을 시뮬레이션하는 기술입니다.',
        metadata: includeMetadata ? { source: 'AI 문서', date: '2023-01-01' } : undefined,
        score: 0.95
      },
      {
        id: '2',
        content: '기계 학습은 AI의 하위 분야로, 데이터로부터 학습하는 알고리즘을 포함합니다.',
        metadata: includeMetadata ? { source: 'ML 문서', date: '2023-02-15' } : undefined,
        score: 0.85
      },
      {
        id: '3',
        content: '자연어 처리(NLP)는 컴퓨터가 인간의 언어를 이해하고 처리하는 AI 분야입니다.',
        metadata: includeMetadata ? { source: 'NLP 문서', date: '2023-03-20' } : undefined,
        score: 0.75
      }
    ];

    // 임계값과 최대 결과 수에 따라 필터링
    const filteredResults = results
      .filter(doc => doc.score && doc.score >= threshold)
      .slice(0, maxResults);

    const executionTime = Date.now() - startTime;

    logger.log({
      message: `RAG 검색 완료: ${filteredResults.length} 결과 찾음`,
      level: 'info',
      context: { executionTime }
    });

    return {
      query,
      results: filteredResults,
      totalCount: filteredResults.length,
      executionTime
    };
  } catch (error) {
    logger.error({
      message: `RAG 검색 실패: ${error.message}`,
      error,
      context: { query, options }
    });

    throw new Error(`RAG 검색 실패: ${error.message}`);
  }
}

/**
 * 검색 결과를 기반으로 콘텐츠를 생성합니다
 * @param query 사용자 쿼리
 * @param searchResults 검색 결과
 * @returns 생성된 콘텐츠와 인용 정보
 */
export async function generateFromResults(
  query: string,
  searchResults: RagSearchResult
): Promise<RagGenerationResult> {
  try {
    logger.log({
      message: `RAG 콘텐츠 생성 시작: "${query}"`,
      level: 'info',
      context: { resultCount: searchResults.results.length }
    });

    // 여기서는 단순한 더미 응답을 반환합니다
    // 실제 구현에서는 여기서 LLM을 호출하여 검색 결과를 바탕으로 콘텐츠를 생성합니다
    const generatedContent = `
    ${query}에 대한 답변입니다:
    
    ${searchResults.results[0]?.content || '관련 정보를 찾을 수 없습니다.'} 
    ${searchResults.results[1]?.content || ''} 
    ${searchResults.results[2]?.content || ''}
    
    결론적으로, 이는 중요한 AI 개념입니다.
    `;

    return {
      sourceQuery: query,
      generatedContent: generatedContent.trim(),
      citations: searchResults.results,
      confidence: 0.85
    };
  } catch (error) {
    logger.error({
      message: `RAG 콘텐츠 생성 실패: ${error.message}`,
      error,
      context: { query, resultCount: searchResults.results.length }
    });

    throw new Error(`RAG 콘텐츠 생성 실패: ${error.message}`);
  }
}

/**
 * 키워드를 해당 카테고리로 분류합니다.
 * @param keyword 분류할 키워드
 * @returns 키워드 카테고리
 */
export function categorizeKeyword(keyword: string): KeywordCategory {
  if (!keyword) return '일반';
  
  const normalizedKeyword = keyword.toLowerCase().trim();
  
  // 카테고리별 키워드 목록과 일치하는지 확인
  const categoryKeywords: Record<KeywordCategory, string[]> = {
    'AI 기술': ['AI', 'GPT', '인공지능', '머신러닝', '딥러닝', 'Claude', 'LLM', '생성형 AI'],
    '디지털 마케팅': ['마케팅', 'SEO', '디지털 마케팅', '소셜미디어', '광고', '콘텐츠 마케팅'],
    '앱 개발': ['앱 개발', '웹 개발', '프로그래밍', '모바일 앱', '웹사이트', '프론트엔드', '백엔드'],
    '교육/학습': ['교육', '학습', '강의', '온라인 강의', 'e러닝', '학교', '대학'],
    '건강/의료': ['건강', '의료', '병원', '운동', '다이어트', '건강관리', '헬스케어'],
    '금융/투자': ['금융', '투자', '주식', '경제', '재테크', '자산관리', '암호화폐'],
    '3D 모델링/AI': ['3D', '모델링', '렌더링', 'CAD', 'Blender', 'Unity', 'Unreal'],
    '일반': []
  };
  
  // 각 카테고리의 키워드 목록과 일치하는지 확인
  for (const [category, keywords] of Object.entries(categoryKeywords)) {
    if (keywords.some(kw => normalizedKeyword.includes(kw.toLowerCase()))) {
      return category as KeywordCategory;
    }
  }
  
  // 일치하는 카테고리가 없으면 '일반' 반환
  return '일반';
}

/**
 * 키워드 배열로부터 마크다운 형식의 분석 결과를 생성합니다.
 * @param keywords 분석할 키워드 배열
 * @param preferences 사용자 설정
 * @returns 마크다운 형식의 분석 결과
 */
export function generateKeywordAnalysis(
  keywords: string[],
  preferences: UserPreferences = {}
): string {
  try {
    // 빈 배열 체크
    if (!keywords || keywords.length === 0) {
      // 언어 설정에 따른 메시지 반환
      return preferences.language === 'en' ? 'No keywords to analyze.' : '분석할 키워드가 없습니다.';
    }
    
    // 기본 설정
    const language = preferences.language || 'ko';
    const insightCount = preferences.insightCount || 3;
    const strategyCount = preferences.strategyCount || 3;
    
    // 키워드 중 첫 번째 것을 대표 키워드로 사용
    const mainKeyword = keywords[0];
    
    // 키워드 카테고리 결정
    let category = categorizeKeyword(mainKeyword);
    
    // 산업 분야가 지정된 경우 카테고리 오버라이드
    if (preferences.industry) {
      const industry = preferences.industry.toLowerCase();
      if (industry.includes('market') || industry.includes('마케팅')) {
        category = '디지털 마케팅';
      } else if (industry.includes('dev') || industry.includes('개발')) {
        category = '앱 개발';
      } else if (industry.includes('ai') || industry.includes('인공지능')) {
        category = 'AI 기술';
      } else if (industry.includes('edu') || industry.includes('교육')) {
        category = '교육/학습';
      } else if (industry.includes('health') || industry.includes('건강') || industry.includes('의료')) {
        category = '건강/의료';
      } else if (industry.includes('finance') || industry.includes('금융') || industry.includes('투자')) {
        category = '금융/투자';
      }
    }
    
    // 언어별 제목과 섹션 이름
    const title = language === 'en' ? `## ${mainKeyword} Keyword Analysis` : `## ${mainKeyword} 키워드 분석`;
    const insightTitle = language === 'en' ? '### Key Insights' : '### 주요 인사이트';
    const strategyTitle = language === 'en' ? '### Content Strategy' : '### 콘텐츠 제작 전략';

    // 인사이트 생성
    const insights = [];
    const insightsByCategory = {
      'AI 기술': [
        language === 'en'
          ? 'AI technology continues to evolve rapidly, with increasing adoption across industries'
          : 'AI 기술은 계속 빠르게 발전하고 있으며, 다양한 산업에서 도입이 증가하고 있습니다',
        language === 'en'
          ? 'Large language models are becoming more accessible to businesses of all sizes'
          : '대규모 언어 모델은 모든 규모의 기업에서 더 쉽게 접근할 수 있게 되었습니다',
        language === 'en'
          ? 'Users are increasingly concerned about ethical AI and data privacy issues'
          : '사용자들은 윤리적 AI와 데이터 프라이버시 문제에 대해 더 관심을 가지고 있습니다'
      ],
      '디지털 마케팅': [
        language === 'en'
          ? 'Content marketing delivers 3x more leads than traditional marketing for 62% less cost'
          : '콘텐츠 마케팅은 기존 마케팅보다 62% 적은 비용으로 3배 더 많은 리드를 창출합니다',
        language === 'en'
          ? 'Video content generates 66% more qualified leads per year'
          : '비디오 콘텐츠는 연간 66% 더 많은 유자격 리드를 생성합니다',
        language === 'en'
          ? 'Mobile-first strategies are essential as mobile traffic exceeds 55% of total web traffic'
          : '모바일 트래픽이 전체 웹 트래픽의 55%를 초과하므로 모바일 우선 전략이 필수적입니다'
      ],
      '일반': [
        language === 'en'
          ? 'Content that answers specific questions performs 70% better in search rankings'
          : '특정 질문에 답하는 콘텐츠는 검색 랭킹에서 70% 더 좋은 성과를 보입니다',
        language === 'en'
          ? 'Visual content is processed 60,000x faster than text by the human brain'
          : '시각적 콘텐츠는 인간의 뇌에서 텍스트보다 60,000배 더 빠르게 처리됩니다',
        language === 'en'
          ? 'Consistent publishing schedules increase audience retention by 25-30%'
          : '일관된 출판 일정은 잠재 고객 유지율을 25-30% 증가시킵니다'
      ]
    };
    
    // 해당 카테고리의 인사이트가 있으면 사용, 없으면 일반 카테고리 사용
    const categoryInsights = insightsByCategory[category] || insightsByCategory['일반'];
    for (let i = 0; i < Math.min(insightCount, categoryInsights.length); i++) {
      insights.push(`- ${categoryInsights[i]}`);
    }
    
    // 전략 생성
    const strategies = [];
    const strategiesByCategory = {
      'AI 기술': [
        language === 'en'
          ? 'Create educational content that demystifies complex AI concepts for your audience'
          : '청중을 위해 복잡한 AI 개념을 이해하기 쉽게 설명하는 교육 콘텐츠를 만드세요',
        language === 'en'
          ? 'Develop case studies showcasing practical AI applications in relevant industries'
          : '관련 산업에서 실용적인 AI 응용 사례를 보여주는 사례 연구를 개발하세요',
        language === 'en'
          ? 'Create comparison content between different AI tools and technologies'
          : '다양한 AI 도구와 기술 간의 비교 콘텐츠를 제작하세요'
      ],
      '디지털 마케팅': [
        language === 'en'
          ? 'Develop a comprehensive content calendar targeting high-value keywords in your niche'
          : '틈새 시장의 고가치 키워드를 타겟팅하는 포괄적인 콘텐츠 캘린더를 개발하세요',
        language === 'en'
          ? 'Create video tutorials demonstrating practical marketing techniques with measurable results'
          : '측정 가능한 결과로 실용적인 마케팅 기법을 보여주는 비디오 튜토리얼을 제작하세요',
        language === 'en'
          ? 'Publish data-driven case studies highlighting ROI of digital marketing strategies'
          : '디지털 마케팅 전략의 ROI를 강조하는 데이터 기반 사례 연구를 발행하세요'
      ],
      '일반': [
        language === 'en'
          ? 'Create comprehensive guides addressing common questions in your topic area'
          : '주제 영역의 일반적인 질문을 다루는 포괄적인 가이드를 만드세요',
        language === 'en'
          ? 'Develop visual content like infographics to simplify complex information'
          : '복잡한 정보를 단순화하는 인포그래픽과 같은 시각적 콘텐츠를 개발하세요',
        language === 'en'
          ? 'Establish a consistent publishing schedule to build audience expectations'
          : '잠재 고객 기대를 구축하기 위한 일관된 출판 일정을 수립하세요'
      ]
    };
    
    // 해당 카테고리의 전략이 있으면 사용, 없으면 일반 카테고리 사용
    const categoryStrategies = strategiesByCategory[category] || strategiesByCategory['일반'];
    for (let i = 0; i < Math.min(strategyCount, categoryStrategies.length); i++) {
      strategies.push(`${i + 1}. ${categoryStrategies[i]}`);
    }
    
    // 최종 마크다운 조합
    const analysisText = [
      title,
      '',
      insightTitle,
      ...insights,
      '',
      strategyTitle,
      ...strategies
    ].join('\n');
    
    return analysisText;
  } catch (error) {
    logger.error({
      message: '키워드 분석 중 오류 발생',
      error: error as Error,
      context: { keywords }
    });

    // 오류 메시지 (언어에 맞게)
    return preferences.language === 'en'
      ? 'An error occurred during keyword analysis. Please try again.'
      : '키워드 분석 중 오류가 발생했습니다. 다시 시도해 주세요.';
  }
}

// 인라인 RAG 유틸리티 기본 내보내기
export default {
  search: ragSearch,
  generate: generateFromResults,
  generateKeywordAnalysis,
  categorizeKeyword
}; 